gridftp/gsi-wuftpd/source/doc/misc/wu_fnmatch.3
.\"
.\" Copyright (c) 1999 WU-FTPD Development Group. 
.\" All rights reserved.
.\" 
.\" Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994 
.\" The Regents of the University of California.  Portions Copyright (c) 
.\" 1993, 1994 Washington University in Saint Louis.  Portions Copyright 
.\" (c) 1996, 1998 Berkeley Software Design, Inc.  Portions Copyright (c) 
.\" 1998 Sendmail, Inc.  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric 
.\" P. Allman.  Portions Copyright (c) 1989 Massachusetts Institute of 
.\" Technology.  Portions Copyright (c) 1997 Stan Barber.  Portions 
.\" Copyright (C) 1991, 1992, 1993, 1994, 1995, 1996, 1997 Free Software 
.\" Foundation, Inc.  Portions Copyright (c) 1997 Kent Landfield. 
.\"
.\" Use and distribution of this software and its source code are governed 
.\" by the terms and conditions of the WU-FTPD Software License ("LICENSE"). 
.\"
.\"	$Id$
.\"
.TH WU_FNMATCH 3 "July 28, 1999"
.UC 7
.SH NAME
wu_fnmatch \- match filename or pathname
.SH SYNOPSIS
.nf
#include "wu_fnmatch.h"

int wu_fnmatch(const char *pattern, const char *string, int flags);
.fi
.SH DESCRIPTION
.I wu_fnmatch
matches patterns according to the rules used by the shell.
It checks the string specified by the
.I string
argument to see if it matches the pattern specified by the
.I pattern
argument.
.PP
The
.I flags
argument modifies the interpretation of
.I pattern
and
.I string.
The value of
.I flags
is the bitwise inclusive OR of any of the following
constants, which are defined in the include file
.IR wu_fnmatch.h .
.TP
FNM_PATHNAME
Slash characters in
.I string
must be explicitly matched by slashes in
.IR pattern .
If this flag is not set, then slashes are treated as regular characters.
.TP
FNM_NOESCAPE
Every occurrence of a backslash (``\e'') followed by a character in
.I pattern
is replaced by that character.
This is done to negate any special meaning for the character.
If this flag is set, then backslashes are treated as regular characters.
.TP
FNM_PERIOD
If this flag is set, then a leading period in
.I string
must be explicity matched by a period in
.IR pattern .
A period is considered leading if it is the first character in
.IR string ,
or if both
.T FNM_PATHNAME
is set and the period immediately follows a slash.
.TP
FNM_LEADING_DIR
If this flag is set, if the entire
.I pattern
has been matched, and there are more characters remaining in the
.IR string ,
and either the last character matched was a slash, or the next character in the
.I string
is a slash, then consider the entire
.I string
matched.
.TP
FNM_CASEFOLD
Normally, upper case and lower case letters are distinct.  This flag causes
them to be considered equal.
.SH RETURNS
If
.I string
matches the pattern specified by
.IR pattern ,
then
.I wu_fnmatch
returns zero.
Otherwise,
.I wu_fnmatch
returns FNM_NOMATCH.
.SH SEE ALSO
sh(1), glob(3), wordexp(3), regexp(3)
.SH BUGS
{} alternatives are not handled.


ChangeLog
2000-10-10  Von Welch  <vwelch@ncsa.uiuc.edu>

	* myproxy_log.c: Removed my_vsnprintf() as it is now in
	string_funcs.c

	* verror.c: Removed my_vsnprintf() as it is now in
	string_funcs.c
	
	* string_funcs.c, string_funcs.h: Created these files.

	* myproxy_destroy.c (main): Bug fix - use created proxy file for
	authentication and not default proxy.
	(main): Bug fix - passphrase was not being correctly read.

2000-10-09  Von Welch  <vwelch@ncsa.uiuc.edu>

	* verror.c (my_vsnprintf): Fixed inverted logic in checking of
	HAVE_VSNPRINTF
	
2000-10-04  Von Welch  <vwelch@ncsa.uiuc.edu>

	* myproxy_server.c (main): Print strerror() when failure occurs
	opening configuration file.

	* myproxy.c(parse_command): Fixed command_value type.

2000-10-03  Von Welch  <vwelch@ncsa.uiuc.edu>

	* ssl_utils.c: Fixes for openssl build with PEM callback
	differences (openssl has a user data parameter while ssleay does
	not).

	* ChangeLog: Created after 0.2alpha1 release



gridftp/gsi-wuftpd/source/src/wu_fnmatch.c
/****************************************************************************    
  Copyright (c) 1999 WU-FTPD Development Group.  
  All rights reserved.
   
  Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994  
    The Regents of the University of California. 
  Portions Copyright (c) 1993, 1994 Washington University in Saint Louis.  
  Portions Copyright (c) 1996, 1998 Berkeley Software Design, Inc.  
  Portions Copyright (c) 1989 Massachusetts Institute of Technology.  
  Portions Copyright (c) 1998 Sendmail, Inc.  
  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric P.  Allman.  
  Portions Copyright (c) 1997 by Stan Barber.  
  Portions Copyright (c) 1997 by Kent Landfield.  
  Portions Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997  
    Free Software Foundation, Inc.    
   
  Use and distribution of this software and its source code are governed   
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").  
   
  If you did not receive a copy of the license, it may be obtained online  
  at http://www.wu-ftpd.org/license.html.  
   
  $Id$  
   
****************************************************************************/
/*
 * Function fnmatch() as specified in POSIX 1003.2-1992, section B.6.
 * Compares a filename or pathname to a pattern.
 */

#include <ctype.h>
#include <stddef.h>
#include <stdio.h>
#include <string.h>

typedef int boolean;
#define FALSE 0
#define TRUE  1

#include "wu_fnmatch.h"

#define	EOS '\0'

static const char *rangematch(const char *pattern, const char *string, int flags)
{
/*
 * A bracket expression starting with an unquoted circumflex character
 * produces unspecified results (IEEE 1003.2-1992, 3.13.2).  This
 * implementation treats it like '!', for consistency with the regular
 * expression syntax.  J.T. Conklin (conklin@ngai.kaleida.com)
 */
    char test = *string;
    boolean negate = ((*pattern == '!') || (*pattern == '^'));
    boolean ok = FALSE;
    if (negate)
	++pattern;
    if (flags & FNM_CASEFOLD)
	test = tolower((unsigned char) test);
    while (*pattern != ']') {
	char c = *pattern++;
	if ((c == '\\') && !(flags & FNM_NOESCAPE))
	    c = *pattern++;
	if (c == EOS)
	    return (NULL);
	if (flags & FNM_CASEFOLD)
	    c = tolower((unsigned char) c);
	if (*pattern == '-') {
	    char c2 = pattern[1];
	    if ((c2 != EOS)
		&& (c2 != ']')) {
		pattern += 2;
		if ((c2 == '\\') && !(flags & FNM_NOESCAPE))
		    c2 = *pattern++;
		if (c2 == EOS)
		    return (NULL);
		if (flags & FNM_CASEFOLD)
		    c2 = tolower((unsigned char) c2);
		/* this is a hack */
		if ((c <= test) && (test <= c2))
		    ok = TRUE;
	    }
	    else if (c == test)
		ok = TRUE;
	}
	else if (c == test)
	    ok = TRUE;
    }
    return ((ok == negate) ? NULL : pattern);
}

int wu_fnmatch(const char *pattern, const char *string, int flags)
{
    const char *stringstart = string;
    while (TRUE) {
	char test;
	char c = *pattern++;
	switch (c) {
	case EOS:
#ifdef FNM_LEADING_DIR
	    if ((flags & FNM_LEADING_DIR)
		&& (*string == '/'))
		return (0);
	    /*
	     * WU-FTPD extension/correction.
	     *
	     * If the pattern ended with a '/', and we're doing
	     * FNM_PATHNAME matching, consider it a match if the
	     * previous string character was a '/' and the current
	     * is not a '/'.
	     */
	    if ((flags & FNM_LEADING_DIR)
		&& (string != stringstart)
		&& (flags & FNM_PATHNAME)
		&& (*(string - 1) == '/'))
		return (0);
#endif
	    return ((*string == EOS) ? 0 : FNM_NOMATCH);
	case '?':
	    if (*string == EOS)
		return (FNM_NOMATCH);
	    if ((*string == '/')
		&& (flags & FNM_PATHNAME))
		return (FNM_NOMATCH);
	    if ((*string == '.')
		&& (flags & FNM_PERIOD)
		&& ((string == stringstart)
		    || ((flags & FNM_PATHNAME)
			&& (*(string - 1) == '/'))))
		return (FNM_NOMATCH);
	    ++string;
	    break;
	case '*':
	    c = *pattern;
	    while (c == '*')
		c = *++pattern;
	    if ((*string == '.')
		&& (flags & FNM_PERIOD)
		&& ((string == stringstart)
		    || ((flags & FNM_PATHNAME)
			&& (*(string - 1) == '/'))))
		return (FNM_NOMATCH);
	    /* Optimize for pattern with * at end or before /. */
	    if (c == EOS)
		if (flags & FNM_PATHNAME) {
#ifdef FNM_LEADING_DIR
		    if (flags & FNM_LEADING_DIR)
			return (0);
#endif
		    return ((strchr(string, '/') == NULL) ? 0 : FNM_NOMATCH);
		}
		else
		    return (0);
	    else if ((c == '/')
		     && (flags & FNM_PATHNAME)) {
		string = strchr(string, '/');
		if (string == NULL)
		    return (FNM_NOMATCH);
		break;
	    }
	    /* General case, use recursion. */
	    for (test = *string; test != EOS; test = *++string) {
		if (!wu_fnmatch(pattern, string, (flags & ~FNM_PERIOD)))
		    return (0);
		if ((test == '/')
		    && (flags & FNM_PATHNAME))
		    break;
	    }
	    return (FNM_NOMATCH);
	case '[':
	    if (*string == EOS)
		return (FNM_NOMATCH);
	    if ((*string == '/')
		&& (flags & FNM_PATHNAME))
		return (FNM_NOMATCH);
	    pattern = rangematch(pattern, string, flags);
	    if (pattern == NULL)
		return (FNM_NOMATCH);
	    ++string;
	    break;
	case '\\':
	    if (!(flags & FNM_NOESCAPE)) {
		c = *pattern++;
		if (c == EOS) {
		    c = '\\';
		    --pattern;
		}
	    }
	    /* FALLTHROUGH */
	default:
	    if (c == *string);
#ifdef FNM_CASEFOLD
	    else if ((flags & FNM_CASEFOLD)
		     && (tolower((unsigned char) c) == tolower((unsigned char) *string)));
#endif
	    else
		return (FNM_NOMATCH);
	    string++;
	    break;
	}
    }
/* NOTREACHED */
}
